enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,274 @@
+          backend:
+            service:
+              name: example
+              port:
+                number: 80
+```
+The aws load balancer controller will provision an ALB.
+
+Also, the user can install [External DNS Operator](https://github.com/openshift/enhancements/blob/master/enhancements/dns/externaldns-operator.md) to set up a convenient DNS name for its service (ingress support is coming soon).
+
+### Workflow Description
+TBD
+
+### API Extensions
+N/A - no api extensions are being introduced in this proposal
+
+### Implementation Details/Notes/Constraints
+
+The name of the new capability is `ingress`. It includes the cluster ingress operator and all its APIs listed below.
+The implementation of the new capability follows the instructions of [how to implement a new capability](https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).
+
+#### HyperShift
+
+HyperShift doesn't pass all the ingress operator's payload to the cluster version operator.
+Certain manifests are filtered out, most importantly: the ingress operator's deployment.
+The ingress operator is considered as part of the hosted control plane, that is a resource not managed by the user hence deployed on a different cluster.
+Theoretically this may question the need for the cluster capability on HyperShift. However:
+- the cluster capability is a common interface for making OpenShift cluster components optional
+- the cluster capability is already implemented for some operators used by HyperShift (e.g. the console operator)
+- the capability provides a clear API which can be used by any party to understand whether the ingress is enabled
+- most importantly: [the ingress operator's APIs](#ingress-operator-apis) are still created by the cluster version operator
+
+#### Standalone OpenShift
+
+Making the ingress operator optional for the standalone OpenShift is a bigger scope than for HyperShift.
+The changes impact the components which are not used (or used differently) in HyperShift like the installer, the authentication and monitoring operators. To prevent the scope creep for [RFE-3395](https://issues.redhat.com/browse/RFE-3395), the installer should forbid the installation without the ingress capability in [the baseline or additional capabilities of the install config](https://github.com/openshift/installer/blob/master/pkg/types/installconfig.go#L479-L491).
+No change to the capability API is needed, a simple `if` can ensure this pre-requisite.
+
+#### Ingress operator APIs
+
+The following APIs are not available if the ingress capability is disabled:
+- CRDs which are part of the cluster operator's manifest payload:
+  - [IngressController](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition.yaml)
+  - [DNSRecord](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition-internal.yaml)
```

Comment:
Wasn't there previous work for user-managed DNS?  Maybe this deserves to be its own capability, also disabled on HyperShift?

Reply:
You mean [Configurable DNS Management for LoadBalancerService Ingress Controllers](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/configurable-dns-loadbalancerservice.md)? This was an enhancement for the IngressController API to have an ability to disable the DNS management for a given ingress controller.
Overall, I agree, the DNS management stands out and seems to be artificially embedded into the ingress operator due to a lack of a better place maybe?
I would say that I'd rather prefer the ExternalDNS Operator (recently added to the OperatorHub) to take over this functionality and remove this part form the ingress operator completely rather then creating a dedicated  capability for it.

Reply:
The ingress operator manages DNS records for ingresscontrollers and gateways.  The ingress operator defines and uses the DNSRecord CRD in the implementation of the operator's DNS management.  This CRD is an internal implementation detail of the ingress operator and isn't supported for any other use.  

It would be nice if the ingress operator used external-dns instead, but we've been saying that for quite some time: https://github.com/openshift/cluster-ingress-operator/pull/59#issue-379243801, https://github.com/openshift/cluster-ingress-operator/issues/77.  * grin *.  

I don't understand why you would want separate capabilities for ingress and for DNS management for ingresscontrollers though.